crates/apub_receive/src/http/inbox_enums.rs

```diff
@@ -0,0 +1,140 @@
+use crate::activities::{
+  comment::{
+    create::CreateComment,
+    delete::DeleteComment,
+    dislike::DislikeComment,
+    like::LikeComment,
+    remove::RemoveComment,
+    undo_delete::UndoDeleteComment,
+    undo_dislike::UndoDislikeComment,
+    undo_like::UndoLikeComment,
+    undo_remove::UndoRemoveComment,
+    update::UpdateComment,
+  },
+  community::{
+    add_mod::AddMod,
+    announce::AnnounceActivity,
+    block_user::BlockUserFromCommunity,
+    delete::DeleteCommunity,
+    remove::RemoveCommunity,
+    remove_mod::RemoveMod,
+    undo_block_user::UndoBlockUserFromCommunity,
+    undo_delete::UndoDeleteCommunity,
+    undo_remove::UndoRemoveCommunity,
+    update::UpdateCommunity,
+  },
+  following::{accept::AcceptFollowCommunity, follow::FollowCommunity, undo::UndoFollowCommunity},
+  post::{
+    create::CreatePost,
+    delete::DeletePost,
+    dislike::DislikePost,
+    like::LikePost,
+    remove::RemovePost,
+    undo_delete::UndoDeletePost,
+    undo_dislike::UndoDislikePost,
+    undo_like::UndoLikePost,
+    undo_remove::UndoRemovePost,
+    update::UpdatePost,
+  },
+  private_message::{
+    create::CreatePrivateMessage,
+    delete::DeletePrivateMessage,
+    undo_delete::UndoDeletePrivateMessage,
+    update::UpdatePrivateMessage,
+  },
+};
+use lemmy_apub_lib::{ActivityCommonFields, ActivityHandlerNew};
+use lemmy_utils::LemmyError;
+use lemmy_websocket::LemmyContext;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Deserialize, Serialize, ActivityHandlerNew)]
+#[serde(untagged)]
+pub enum PersonInboxActivities {
+  AcceptFollowCommunity(AcceptFollowCommunity),
+  CreatePrivateMessage(CreatePrivateMessage),
+  UpdatePrivateMessage(UpdatePrivateMessage),
+  DeletePrivateMessage(DeletePrivateMessage),
+  UndoDeletePrivateMessage(UndoDeletePrivateMessage),
+  AnnounceActivity(Box<AnnounceActivity>),
+}
```

Comment:
We should ask on stackoverflow the best way to handle deserializing a group of enums all with their own type. Something tells me there's a better way to do it than `serde_json::from_str::<T>(activity)?;`

Logically it seems like `PersonInboxActivity` should be a parent trait which these structs implement.

Reply:
you could look at erased-serde, which allows deserializing into a `dyn` type: https://github.com/dtolnay/erased-serde

Reply:
Interesting, I might give that a try. I'm curious how compilation speed and runtime performance would theoretically compare to the match-based implementation.

Reply:
That library seems quite low-level. There is [typetag](https://github.com/dtolnay/typetag) which is based on it, and seems almost perfect for our use case. The problem is that it doesnt support the untagged enum type which is necessary to parse apub activities. There is [this issue](https://github.com/dtolnay/typetag/issues/12) about it, which suggests that the performance of deserializing into an untagged enum would be quite bad, but I think its definitely worth the simplified code.

Reply:
Might be worth opening an issue for using typetag just so we don't forget.